File: src/main/java/org/modelingvalue/nelumbo/ExistentialQuantifier.java
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
// (C) Copyright 2018-2025 Modeling Value Group B.V. (http://modelingvalue.org)                                        ~
//                                                                                                                     ~
// Licensed under the GNU Lesser General Public License v3.0 (the 'License'). You may not use this file except in      ~
// compliance with the License. You may obtain a copy of the License at: https://choosealicense.com/licenses/lgpl-3.0  ~
// Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on ~
// an 'AS IS' BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the  ~
// specific language governing permissions and limitations under the License.                                          ~
//                                                                                                                     ~
// Maintainers:                                                                                                        ~
//     Wim Bast, Tom Brus                                                                                              ~
//                                                                                                                     ~
// Contributors:                                                                                                       ~
//     Victor Lap                                                                                                      ~
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

package org.modelingvalue.nelumbo;

import java.io.Serial;

import org.modelingvalue.collections.List;
import org.modelingvalue.collections.Set;
import org.modelingvalue.nelumbo.patterns.Functor;

public final class ExistentialQuantifier extends Quantifier {
    @Serial
    private static final long serialVersionUID = 7237451657373739426L;

    private static Functor    FUNCTOR;

    static {
        KnowledgeBase.registerFunctorSetter(ExistentialQuantifier.class, f -> FUNCTOR = f);
    }

    public ExistentialQuantifier(Functor functor, List<AstElement> elements, Object[] args) {
        super(functor, elements, args);
    }

    protected ExistentialQuantifier(List<AstElement> elements, List<Variable> localVars, Predicate predicate) {
        super(FUNCTOR, elements, localVars, predicate);
    }

    private ExistentialQuantifier(Object[] args, ExistentialQuantifier declaration) {
        super(args, declaration);
    }

    @Override
    protected ExistentialQuantifier struct(Object[] array, Predicate declaration) {
        return new ExistentialQuantifier(array, (ExistentialQuantifier) declaration);
    }

    @Override
    protected InferResult resolve(InferContext context, InferResult predResult) {
        List<Variable> localVars = localVars();
        Set<Predicate> facts = Set.of(), falsehoods = Set.of();
        for (Predicate predFact : predResult.facts()) {
            Predicate fact = setBinding(predFact.getBinding().removeAllKey(localVars));
            facts = facts.add(fact);
        }
        for (Predicate predFalsehood : predResult.falsehoods()) {
            Predicate falsehood = setBinding(predFalsehood.getBinding().removeAllKey(localVars));
            if (!facts.contains(falsehood)) {
                falsehoods = falsehoods.add(falsehood);
            }
        }
        boolean completeFacts = predResult.completeFacts();
        boolean completeFalsehoods = predResult.completeFalsehoods();
        if (completeFacts && facts.isEmpty() && completeFalsehoods && falsehoods.isEmpty() && isFullyBound()) {
            falsehoods = falsehoods.add(this);
        }
        return InferResult.of(facts, completeFacts, falsehoods, completeFalsehoods, predResult.cycles());
    }

}

File: web-ts/src/logic/ExistentialQuantifier.ts
/**
 * ExistentialQuantifier - exists quantifier.
 * @JAVA_REF org.modelingvalue.nelumbo.logic.ExistentialQuantifier
 */

import { List, Set } from 'immutable';
import type { AstElement } from '../AstElement';
import { Variable } from '../Variable';
import { Node } from '../Node';
import type { Functor } from '../patterns/Functor';
import { Predicate } from './Predicate';
import { Quantifier } from './Quantifier';
import { InferResult } from '../InferResult';
import type { InferContext } from '../InferContext';

/**
 * ExistentialQuantifier - "there exists" quantification.
 * @JAVA_REF org.modelingvalue.nelumbo.logic.ExistentialQuantifier
 */
export class ExistentialQuantifier extends Quantifier {
  // @JAVA_REF org.modelingvalue.nelumbo.logic.ExistentialQuantifier#FUNCTOR
  static FUNCTOR: Functor | null = null;

  constructor(functor: Functor, elements: List<AstElement>, ...args: unknown[]) {
    super(functor, elements, ...args);
  }

  protected static fromDataExistential(data: unknown[], declaration?: Node): ExistentialQuantifier {
    const eq = Object.create(ExistentialQuantifier.prototype) as ExistentialQuantifier;
    (eq as unknown as { _data: unknown[] })._data = data;
    (eq as unknown as { _declaration: Node })._declaration = declaration ?? eq;
    (eq as any)._binding = null;
    (eq as any)._hashCodeCached = false;
    (eq as any)._hashCode = 0;
    (eq as any)._nrOfUnbound = -1;
    return eq;
  }

  protected override struct(data: unknown[], declaration?: Node): ExistentialQuantifier {
    return ExistentialQuantifier.fromDataExistential(data, declaration ?? this.declaration());
  }

  // @JAVA_REF org.modelingvalue.nelumbo.logic.ExistentialQuantifier#resolve(InferContext, InferResult)
  protected override resolveWithResult(_context: InferContext, predResult: InferResult): InferResult {
    const localVars = this.localVars();
    let facts = Set<Predicate>();
    let falsehoods = Set<Predicate>();
    let completeFacts = true;
    let completeFalsehoods = true;

    for (const predFact of predResult.facts()) {
      const binding = predFact.getBinding();
      const cleanBinding = binding !== null ? binding.deleteAll(localVars) : binding;
      const fact = this.setBinding(cleanBinding!);
      if (fact.isFullyBound()) {
        facts = facts.add(fact);
      } else {
        completeFacts = false;
      }
    }

    for (const predFalsehood of predResult.falsehoods()) {
      const binding = predFalsehood.getBinding();
      const cleanBinding = binding !== null ? binding.deleteAll(localVars) : binding;
      const falsehood = this.setBinding(cleanBinding!);
      if (!facts.contains(falsehood)) {
        if (falsehood.isFullyBound()) {
          falsehoods = falsehoods.add(falsehood);
        } else {
          completeFalsehoods = false;
        }
      }
    }

    if (!this.isFullyBound()) {
      if (!predResult.completeFacts()) {
        completeFacts = false;
      }
      if (!predResult.completeFalsehoods()) {
        completeFalsehoods = false;
      }
    } else if (falsehoods.isEmpty() && facts.isEmpty()) {
      falsehoods = falsehoods.add(this as Predicate);
    }

    return InferResult.of(facts, completeFacts, falsehoods, completeFalsehoods, predResult.cycles());
  }
}
